### ISSUE TITLE
New rule proposal: `no-useless-constructor`

### ISSUE THREAD

**Issue by MIreland** [oxc-project/tsgolint#754]
## Summary

This is a proposal to implement the TypeScript extension of ESLint's [`no-useless-constructor`](https://eslint.org/docs/latest/rules/no-useless-constructor) rule.

> **Note:** This issue was opened by an AI assistant (Claude, by Anthropic) on behalf of a user who identified this gap while auditing typescript-eslint extension rule coverage in tsgolint.

---

## Rule description

[`@typescript-eslint/no-useless-constructor`](https://typescript-eslint.io/rules/no-useless-constructor) extends the base ESLint rule to understand TypeScript-specific constructor patterns that make a constructor useless:

- **Parameter properties** — a constructor whose only body is parameter property declarations (e.g. `constructor(private x: string) {}`) is not useless in TypeScript, and the TS version correctly avoids false positives here
- **Visibility modifiers** — a constructor that only changes access level (e.g. `protected constructor() { super(); }`) may or may not be useless; the TS version flags these (with a known caveat: type information would be needed to fully resolve this)
- **Empty constructors with no superclass** — the base rule handles these; the TS extension ensures TypeScript-specific inheritance patterns don't cause false positives

Without this rule (or with the base ESLint rule enabled instead), TypeScript projects get either false positives (base rule flags valid TS patterns) or no coverage at all.

## Example

```ts
// ✅ OK — parameter property, NOT useless
class Foo {
  constructor(private x: string) {}
}

// ❌ Error — useless, only calls super with same args
class Bar extends Foo {
  constructor(x: string) {
    super(x);
  }
}

// ❌ Error — empty constructor, no superclass
class Baz {
  constructor() {}
}
```

## Motivation

This is one of 18 typescript-eslint [extension rules](https://typescript-eslint.io/rules/?=extension) (🧱) not yet covered by tsgolint. It's a commonly-enabled rule in TypeScript strict configs (`plugin:@typescript-eslint/strict`) and is a direct replacement for the base ESLint rule — users migrating from ESLint would expect it to be available.

## Configuration

No options required. When enabling this rule, users should disable the base ESLint rule:

```json
{
  "no-useless-constructor": "off",
  "@typescript-eslint/no-useless-constructor": "error"
}
```

## References

- typescript-eslint rule docs: https://typescript-eslint.io/rules/no-useless-constructor
- Base ESLint rule: https://eslint.org/docs/latest/rules/no-useless-constructor

---

## Related coverage gaps

While auditing [typescript-eslint extension rules](https://typescript-eslint.io/rules/?=extension) (🧱) against tsgolint, the following rules are also missing. Listing here for visibility — happy to open separate issues if useful:

| Rule | Notes |
|------|-------|
| `no-unused-expressions` | partial type-awareness |
| `class-methods-use-this` | |
| `default-param-last` | |
| `init-declarations` | |
| `max-params` | |
| `no-array-constructor` | |
| `no-dupe-class-members` | |
| `no-empty-function` | |
| `no-invalid-this` | |
| `no-loop-func` | |
| `no-loss-of-precision` | |
| `no-magic-numbers` | |
| `no-redeclare` | |
| `no-restricted-imports` | |
| `no-shadow` | |
| `no-unused-vars` | |
| `no-use-before-define` | |

(`prefer-destructuring` is already tracked as planned.)

**Comment by connorshea** [comment]
@MIreland I think there's a misunderstanding here with regards to how these rules are implemented in Oxlint/tsgolint. For one, nearly all of the rules in the list here have no type-aware logic. As such, they would be implemented in Oxlint, there's no reason to add the overhead of tsgolint for these rules. If they are not indicated in typescript-eslint docs as being type-aware, they should be implemented in Oxlint.

And then, in most cases where the TypeScript rule extends an ESLint core rule, we implement the full behavior from both libraries (where possible) in the `eslint/` namespaced rule in Oxlint. The list of those is here (though not all are implemented yet): https://github.com/oxc-project/oxc/blob/f78e6df01a5e73d0be60890a8ae7cb28ae241220/crates/oxc_linter/src/utils/mod.rs#L90

I think we could improve the documentation to communicate this better, but most of the logic for these extension rules _is_ already present in Oxlint. You can see in [the tests for some of the extension rules ported to Oxlint](https://github.com/oxc-project/oxc/blob/f78e6df01a5e73d0be60890a8ae7cb28ae241220/crates/oxc_linter/src/rules/eslint/no_dupe_class_members.rs) that we ported the tests from the original ESLint rule _and_ the typescript-eslint rule.

The original ESLint rules in ESLint itself have _also_ been upgraded over the years to support TypeScript syntax, see https://github.com/typescript-eslint/typescript-eslint/issues/11014 for an issue in typescript-eslint where they discuss/consider deprecating some of these extension rules entirely now that the original ESLint rules cover the same use-cases.

All that to say, I think nearly all of these extension rules are already covered in the Oxlint rules, including partially `no-useless-constructor` (although there are [a handful of tests](https://github.com/oxc-project/oxc/blob/f78e6df01a5e73d0be60890a8ae7cb28ae241220/crates/oxc_linter/src/rules/eslint/no_useless_constructor.rs#L341) that did not get enabled due to requiring type-awareness in some cases?).

**Comment by connorshea** [comment]
Also, re-reading that I hope my comment doesn't come off as too harsh 😅 I didn't intend it that way, just wanted to make clear that these rules do exist and are already implemented. The oxlint-migrate script even handles this for you when you have these extension rules enabled in your original ESLint config, porting them to the shared/merged Oxlint rules which include the behavior from the original and extension rules.

We should probably consider how we can improve the docs here to communicate this, though.

**Comment by MIreland** [comment]
@connorshea I deeply appreciate you spending the time to write all that up in response to my clearly-ai-written bug report 🙈

I actually did pick up on the merged behavior eventually - although I agree it isn't the most cleanly surfaced.

I'm specifically calling this rule out because the [oxlint docs themselves do:](https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-useless-constructor.html#eslint-no-useless-constructor)

<img width="709" height="197" alt="Image" src="https://github.com/user-attachments/assets/3b67371d-cb12-4463-b60b-865f500c8969" />

We use aggressive pre-commit auto-fix settings, and when we swapped from eslint to oxlint some folks saw undesirable behavior related to loosening of a type when a "useless" constructor is removed.
